[DmaComposition] Rework logic to enable new composition opportunity

newling opened the pull request:
- slight modification to logic used for compositing DMAs, makes use of global offset difference (please see comments in code)
- more aggressive folding of unit dimensions 
- new unit test in combine_strided_ops.mlir shows the case that was being missed in an end-to-end matmul transpose b with the new pipeline pack-peel-4-level-tiling

yzhang93 commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:
```diff
@@ -511,6 +480,91 @@ LogicalResult moveNpuDmaSyncUsersAfterAncestorInSameBlock(
 }
 
+/// Try to add `offsetToMerge` to one of the offsets in `offsets`.  This
+/// function assumes an effective stride for `offsetToMerge` of 1one so to add
```
I'm not sure what it should be. 1one -> one?

newling replied:
eleven ;) 
I'll fix, thanks.

yzhang93 reviewed (commented):
LGTM. Some minor comments.

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp:
```diff
@@ -326,6 +332,9 @@ TEST_F(AccessPatternCombinationTest, CombineAccessPatterns) {
   checkCombineAccessPatterns({0, 0}, {16, 32}, {16, 1}, {1, 0, 0}, {3, 16, 32},
                              {64, 16, 1}, {0, 0, 0}, {4, 16, 32}, {64, 16, 1},
                              3);
+  // stride matching
```
Could you add a few more test cases for this case with more dimensions and with non-zero offsets for A? E.g. something like this:
```
checkCombineAccessPatterns({4, 0}, {8, 16}, {64, 1}, {4 * 64 + 8 * 64, 0}, {1, 16},
                             {1, 1}, {4, 0}, {9, 16}, {64, 1}, 3, true);
```

newling replied:
I've added this and a few others

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:
```diff
@@ -7,201 +7,275 @@
+/// by the access pattern B, can be merged into a single access pattern.
+///
+/// \return global_offset(X) - global_offset(Y).
+std::optional<int64_t> getGlobalOffsetDifference(
```
I think it would be useful to add unit tests for this utility (and other similar utilities).

newling replied:
I've added some in the .cpp test file for the constant cases. For the non-constant cases, I'm in the process of adding in the combine_strided_ops.mlir, need to add some negative cases though (non-constant global offset difference). Tomorrow..

jtuyls reviewed (changes requested):
This is a great refactor with additional combinations being enabled, thanks for the thorough logic and great explanations!

newling commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:
```diff
@@ -7,201 +7,275 @@
+    return true;
+  };
+
+  for (uint32_t i = 0; i < offsetsX.size(); ++i) {
```
That was what I had originally, but then there are missed opportunities where the difference is constant, but the individual global offsets are not. I hit these cases with some of the end-to-end matmuls. 

Something like 
X = offsets:  [%v] sizes: [10] strides: [1]
Y = offsets:  [%v]  sizes: [10] strides: [1]

for example: the global offset of X (and Y) is %v  (i.e. non-constant), but the difference is %v - %v = 0 is constant. So by computing the global difference directly we see we can merge to 

merged = offsets: [0, %v] sizes: [2, 10] strides: [0, 1]. 

Another option could be to actually build a global offset expression with arith ops for X and Y, and then hope to somehow canonicalize the difference between them and hope to get a constant. This is what I had originally, but it didn't work.  

I will add more tests with non-constant offsets illustrating this.

jtuyls replied:
Thanks for the explanation! I think it would be useful to add this to the docstring of the function as well. Can be brief, as a reference for someone touching that function in the future.

newling commented:
> This is a great refactor with additional combinations being enabled, thanks for the thorough logic and great explanations!

Thanks! I think I've addressed all the comments now, let me know if you think it should have more testing etc.

jtuyls commented on compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp:
```diff
@@ -6,15 +6,27 @@
 using namespace mlir;
 using namespace mlir::iree_compiler::AMDAIE;
 
+SmallVector<int64_t> fromOpFoldResults(SmallVector<OpFoldResult> ofrs) {
```
Can't you use `getConstantIntValues`?

newling replied:
Oops second time. Changed

jtuyls reviewed (approved):
LGTM, just one nit above.
